src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs

```diff
@@ -292,23 +292,34 @@ public DependencyContext Build(string[] userRuntimeAssemblies = null)
 
             foreach (var directReference in directAndDependencyReferences)
             {
-                var runtimeLibrary = new RuntimeLibrary(
+                var runtimeLibrary = new ModifiableRuntimeLibrary(new RuntimeLibrary(
                     type: "reference",
                     name: GetReferenceLibraryName(directReference),
                     version: directReference.Version,
                     hash: string.Empty,
-                    runtimeAssemblyGroups: new[] { new RuntimeAssetGroup(string.Empty, new[] { CreateRuntimeFile(directReference.FileName, directReference.FullPath) }) },
-                    nativeLibraryGroups: new RuntimeAssetGroup[] { },
+                    runtimeAssemblyGroups: [new RuntimeAssetGroup(string.Empty, [CreateRuntimeFile(directReference.FileName, directReference.FullPath)])],
+                    nativeLibraryGroups: [],
                     resourceAssemblies: CreateResourceAssemblies(directReference.ResourceAssemblies),
-                    dependencies: Enumerable.Empty<Dependency>(),
+                    dependencies: [],
                     path: null,
                     hashPath: null,
                     runtimeStoreManifestName: null,
-                    serviceable: false);
+                    serviceable: false), []);
 
                 runtimeLibraries.Add(runtimeLibrary);
             }
 
+            /*
+             * We now need to modify runtimeLibraries to eliminate those that don't have any runtime assets. We follow the following steps:
+             * 0. Construct a reverse dependencies list: all runtimeLibraries that depend on this one
+             * 1. If runtimeAssemblyGroups, nativeLibraryGroups, dependencies, and resourceAssemblies are all empty, remove this runtimeLibrary as well as any dependencies on it.
+             * 2. Add all runtimeLibraries to a list of to-be-processed libraries called libraryCandidatesForRemoval
+             * 3. libraryCandidatesForRemoval.Pop() --> if there are no runtimeAssemblyGroups, nativeLibraryGroups, or resourceAssemblies, and either dependencies is empty or all
+             *      dependencies have something else that depends on them, remove it (and from libraryCandidatesForRemoval), adding everything that depends on this to 
+             *      libraryCandidatesForRemoval if it isn't already there
+             * Repeat 3 until libraryCandidatesForRemoval is empty
+             */
+
```

Comment:
Yes, this looks like it should work.

It may make sense to update ModifiableRuntimeLibrary to have lists of both the dependencies and the dependants (the other direction) so that you can use that as the graph of objects to walk.

Instead of a stack you may also want to use something like a HashSet so you can put a candidate for removal in the list for processing without having to process it multiple times if more than one library that gets removed has it as a dependant before you get to processing that library.